# 生信编程直播第三题-hg38每条染色体基因,转录本的分布
## 题目

这一题主要针对，gtf注释文件的探究！

以R完成的版本：[R的bioconductor包TxDb.Hsapiens.UCSC.hg19.knownGene详解](http://www.bio-info-trainee.com/831.html)

人类的基因组注释文件，一般是gtf/gff格式的。参考：http://www.bio-info-trainee.com/782.html
用perl和python，R，shell均可以完成！

- 基础作业：对这个文件 ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.chr.gtf.gz  进行统计，普通人只需要关心第1，3列就好了。
    - 每条染色体基因个数的分布？
    - 所有基因平均有多少个转录本？
    - 所有转录本平均有多个exon和intron？
    - 或者其它一系列你想统计的，你想了解的！

- 高级作业：下载human,rat,mouse,dog,cat,chicken,等十几种物种的gtf文件：http://asia.ensembl.org/info/data/ftp/index.html 然后写好染色体基因,转录本的分布统计函数，转录本外显子个数统计分布函数，对它们批量处理。

- 发挥想象力：
    - 所有基因平均有多少个转录本？
    - 所有转录本平均有多个exon和intron？
    - 如果要比较多个数据库呢？gencode？UCSC？NCBI？
    - 如果把基因分成多个类型呢？protein coding gene，pseudogene，lncRNA还有miRNA的基因？它们的特征又有哪些变化呢？

## 我的答案

1. 首先当然是下载：

```
wget ftp://ftp.ensembl.org/pub/release-87/gtf/homo_sapiens/Homo_sapiens.GRCh38.87.chr.gtf.gz
```

2. `head`看看（gtf格式的详细说明参考：[NGS数据格式之gff/gtf](https://mp.weixin.qq.com/s?__biz=MzUzMTEwODk0Ng==&mid=2247484320&idx=1&sn=f7c9e7ee096f636449aa5137af7b830d&scene=21#wechat_redirect)）

```
zcat Homo_sapiens.GRCh38.87.chr.gtf.gz | head
```

```
#!genome-build GRCh38.p7
#!genome-version GRCh38
#!genome-date 2013-12
#!genome-build-accession NCBI:GCA_000001405.22
#!genebuild-last-updated 2016-06
1	havana	gene	11869	14409	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; havana_gene "OTTHUMG00000000961"; havana_gene_version "2";
1	havana	transcript	11869	14409	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; havana_gene "OTTHUMG00000000961"; havana_gene_version "2"; transcript_name "DDX11L1-002"; transcript_source "havana"; transcript_biotype "processed_transcript"; havana_transcript "OTTHUMT00000362751"; havana_transcript_version "1"; tag "basic"; transcript_support_level "1";
1	havana	exon	11869	12227	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; exon_number "1"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; havana_gene "OTTHUMG00000000961"; havana_gene_version "2"; transcript_name "DDX11L1-002"; transcript_source "havana"; transcript_biotype "processed_transcript"; havana_transcript "OTTHUMT00000362751"; havana_transcript_version "1"; exon_id "ENSE00002234944"; exon_version "1"; tag "basic"; transcript_support_level "1";
1	havana	exon	12613	12721	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; exon_number "2"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; havana_gene "OTTHUMG00000000961"; havana_gene_version "2"; transcript_name "DDX11L1-002"; transcript_source "havana"; transcript_biotype "processed_transcript"; havana_transcript "OTTHUMT00000362751"; havana_transcript_version "1"; exon_id "ENSE00003582793"; exon_version "1"; tag "basic"; transcript_support_level "1";
1	havana	exon	13221	14409	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; transcript_id "ENST00000456328"; transcript_version "2"; exon_number "3"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene"; havana_gene "OTTHUMG00000000961"; havana_gene_version "2"; transcript_name "DDX11L1-002"; transcript_source "havana"; transcript_biotype "processed_transcript"; havana_transcript "OTTHUMT00000362751"; havana_transcript_version "1"; exon_id "ENSE00002312635"; exon_version "1"; tag "basic"; transcript_support_level "1";
```

3. 计算每条染色体的基因个数：

我的思路：首先把前面的注释去掉，然后根据第三列的`gene`来得到基因，用`sort`排列后再用`uniq -c`计数。

```
zcat Homo_sapiens.GRCh38.87.chr.gtf.gz | sed '/^#/d' |cut -f1,3 |  grep 'gene' | sort -n |uniq -c | cut -f1
    37 MT
   2359 X
    523 Y
   5194 1
   3971 2
   3010 3
   2505 4
   2868 5
   2863 6
   2867 7
   2353 8
   2242 9
   2204 10
   3235 11
   2940 12
   1304 13
   2224 14
   2152 15
   2511 16
   2995 17
   1170 18
   2926 19
   1386 20
    835 21
   1318 22
```

4. 如果只考虑protein coding gene：（就是多加了个筛选条件嘛）

```
zcat Homo_sapiens.GRCh38.87.chr.gtf.gz | sed '/^#/d' | awk '{if($3 == "gene")print}' | grep 'protein_coding' | cut -f1 | sort -n |uniq -c
     13 MT
    841 X
     53 Y
   2052 1
   1255 2
   1076 3
    751 4
    876 5
   1045 6
    906 7
    676 8
    781 9
    732 10
   1276 11
   1034 12
    327 13
    623 14
    597 15
    866 16
   1197 17
    270 18
   1470 19
    544 20
    233 21
    438 22
```

5. 基因分类：

我的思路：在筛选得到`gene`的基础上，再正则匹配输出 `gene_biotype`开头的行，因为之后的内容都是以`;`分隔，且 `gene_biotype`的内容在第一列，所以用`awk`的内置变量`$1`取出；之后就是排序，计数，输出

```
zcat Homo_sapiens.GRCh38.87.chr.gtf.gz | sed '/^#/d' | awk '{if($3 == "gene")print}' | grep -o "gene_biotype.*;" |awk 'BEGIN{FS=";"}{print $1}' | sort |uniq -c| awk '{print $1,$3}' |sort -rn
19932 "protein_coding"
10268 "processed_pseudogene"
7533 "lincRNA"
5526 "antisense"
2661 "unprocessed_pseudogene"
2212 "misc_RNA"
1900 "snRNA"
1567 "miRNA"
1048 "TEC"
944 "snoRNA"
908 "sense_intronic"
735 "transcribed_unprocessed_pseudogene"
543 "rRNA"
515 "processed_transcript"
450 "transcribed_processed_pseudogene"
193 "IG_V_pseudogene"
187 "sense_overlapping"
154 "unitary_pseudogene"
145 "IG_V_gene"
108 "TR_V_gene"
79 "TR_J_gene"
60 "transcribed_unitary_pseudogene"
51 "polymorphic_pseudogene"
49 "scaRNA"
37 "IG_D_gene"
30 "TR_V_pseudogene"
30 "3prime_overlapping_ncRNA"
22 "Mt_tRNA"
21 "pseudogene"
18 "IG_J_gene"
14 "IG_C_gene"
9 "IG_C_pseudogene"
8 "ribozyme"
6 "TR_C_gene"
5 "sRNA"
4 "bidirectional_promoter_lncRNA"
4 "TR_J_pseudogene"
4 "TR_D_gene"
3 "non_coding"
3 "IG_J_pseudogene"
2 "Mt_rRNA"
1 "vaultRNA"
1 "scRNA"
1 "macro_lncRNA"
1 "IG_pseudogene"
```

6. 我只完成了基础作业，之后的高级作业会在下次更新给出。

---
:yum: 我还是一个小小白，平时的笔记一定有不少疏漏之处，恳请各位大佬批评指正！
